transformer_engine/common/transpose/cast_transpose.cu

```diff
@@ -6,432 +6,360 @@
-                                              THREADS_PER_WARP;
-  OVec * const my_scratch = reinterpret_cast<OVec*>(scratch) +
-                            (my_id_in_warp + warp_id / n_warps_per_tile * THREADS_PER_WARP) *
-                            (THREADS_PER_WARP + 1);
-
-  IVec in[2][nvec_out];
-  const unsigned int warp_id_in_tile = warp_id % n_warps_per_tile;
-  constexpr unsigned int n_iterations = THREADS_PER_WARP / n_warps_per_tile;
-  OVec out_space[n_iterations][nvec_in];
-
-  const size_t stride = row_length / nvec_in;
-  const size_t output_stride = num_rows / nvec_out;
-  size_t current_stride = warp_id_in_tile * n_iterations * nvec_out * stride;
-  unsigned int my_place = (my_id_in_warp + THREADS_PER_WARP -
-                           warp_id_in_tile * n_iterations) %
-                         THREADS_PER_WARP;
-  CType max = 0;
-  const CType scale = scale_ptr != nullptr ? *scale_ptr : 1;
-#pragma unroll
-  for (unsigned int i = 0; i < nvec_out; ++i) {
-    in[0][i].load_from(my_input_tile, current_stride + my_place + stride * i);
-  }
-#pragma unroll
-  for (unsigned int i = 0; i < n_iterations; ++i) {
-    const size_t current_place = current_stride + my_place;
-    const unsigned int my_place_in = (my_place + THREADS_PER_WARP - 1) % THREADS_PER_WARP;
-    const unsigned int current_in = (i + 1) % 2;
-    if (i < n_iterations - 1) {
-#pragma unroll
-      for (unsigned int j = 0; j < nvec_out; ++j) {
-        in[current_in][j].load_from(my_input_tile,
-                                    current_stride + my_place_in + stride * (nvec_out + j));
-      }
+namespace {
+
+// String with RTC kernel implementation
+#include "string_code_transpose_rtc_cast_transpose_cu.h"
+
+// Hard-coded kernel parameters
+using CType = float;
+constexpr size_t warps_per_tile = 4;
+constexpr size_t block_size = THREADS_PER_WARP * warps_per_tile;
+
+/* Performance heuristics for optimized kernel parameters */
+struct KernelConfig {
+  /** Vector load size */
+  size_t load_size = 0;
+  /** Vector store size to transposed output */
+  size_t store_size = 0;
+
+  /* Whether config is valid */
+  bool valid = false;
+  /* Number of CUDA blocks */
+  size_t num_blocks = 0;
+
+  /* Number of active SMs */
+  size_t active_sm_count = 0;
+  /* Elements per L1 cache load */
+  size_t elements_per_load = 0;
+  /* Elements per L1 cache store to cast output*/
+  size_t elements_per_store_c = 0;
+  /* Elements per L1 cache store to transposed output */
+  size_t elements_per_store_t = 0;
+
+  KernelConfig(size_t row_length,
+               size_t num_rows,
+               size_t itype_size,
+               size_t otype_size,
+               size_t load_size_,
+               size_t store_size_)
+    : load_size{load_size_}
+    , store_size{store_size_} {
+    // Check that tiles are correctly aligned
+    constexpr size_t cache_line_size = 128;
+    if (load_size % itype_size != 0
```

Comment:
Maybe to add a notice that the object has not been fully constructed and might be in inconsistent state?